File: src/test/java/no/nav/bidrag/dokument/CorrelationIdFilterTest.java
package no.nav.bidrag.dokument;

import static no.nav.bidrag.dokument.BidragDokumentLocal.TEST_PROFILE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.Appender;
import no.nav.bidrag.commons.KildesystemIdenfikator;
import no.nav.bidrag.commons.security.service.OidcTokenManager;
import no.nav.bidrag.commons.web.HttpResponse;
import no.nav.bidrag.commons.web.test.HttpHeaderTestRestTemplate;
import no.nav.bidrag.dokument.dto.JournalpostResponse;
import no.nav.bidrag.dokument.service.JournalpostService;
import no.nav.security.token.support.spring.test.EnableMockOAuth2Server;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.ActiveProfiles;

@SpringBootTest(classes = BidragDokumentLocal.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles(TEST_PROFILE)
@DisplayName("CorrelationIdFilter")
@EnableMockOAuth2Server
class CorrelationIdFilterTest {

  @Autowired
  private HttpHeaderTestRestTemplate securedTestRestTemplate;

  @SuppressWarnings("rawtypes")
  @MockBean
  private Appender appenderMock;
  @MockBean
  private JournalpostService journalpostServiceMock;

  @LocalServerPort
  private int port;

  @MockBean
  OidcTokenManager oidcTokenManager;

  @BeforeEach
  @SuppressWarnings("unchecked")
  void mockLogAppender() {
    ch.qos.logback.classic.Logger logger = (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
    when(appenderMock.getName()).thenReturn("MOCK");
    when(appenderMock.isStarted()).thenReturn(true);
    logger.addAppender(appenderMock);
  }

  @Test
  @SuppressWarnings("unchecked")
  @DisplayName("skal logge requests mot applikasjonen")
  void skalLoggeRequestsMotApplikasjonen() {
    when(oidcTokenManager.isValidTokenIssuedByAzure()).thenReturn(false);
    when(oidcTokenManager.fetchTokenAsString()).thenReturn("");
    when(journalpostServiceMock.hentJournalpost(anyString(), any(KildesystemIdenfikator.class)))
        .thenReturn(HttpResponse.from(HttpStatus.I_AM_A_TEAPOT));

    var response = securedTestRestTemplate.exchange(
        "http://localhost:" + port + "/bidrag-dokument/journal/BID-123?saksnummer=777",
        HttpMethod.GET,
        null,
        JournalpostResponse.class
    );

    assertAll(
        () -> assertThat(response).extracting(ResponseEntity::getStatusCode).isEqualTo(HttpStatus.I_AM_A_TEAPOT),
        () -> {
          var loggingEventCaptor = ArgumentCaptor.forClass(ILoggingEvent.class);
          verify(appenderMock, atLeastOnce()).doAppend(loggingEventCaptor.capture());
        }
    );
  }
}

File: src/test/java/no/nav/bidrag/dokument/CorrelationIdFilterTest.kt
package no.nav.bidrag.dokument

import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.Appender
import com.ninjasquad.springmockk.MockkBean
import io.kotest.assertions.assertSoftly
import io.mockk.every
import no.nav.bidrag.commons.util.KildesystemIdenfikator
import no.nav.bidrag.commons.web.HttpResponse
import no.nav.bidrag.commons.web.test.HttpHeaderTestRestTemplate
import no.nav.bidrag.dokument.service.JournalpostService
import no.nav.bidrag.transport.dokument.JournalpostResponse
import no.nav.security.token.support.spring.test.EnableMockOAuth2Server
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mockito
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.http.HttpStatus
import org.springframework.test.context.ActiveProfiles

@SpringBootTest(
    classes = [BidragDokumentTest::class],
    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
)
@ActiveProfiles(BidragDokumentTest.TEST_PROFILE, "mock-security")
@DisplayName("CorrelationIdFilter")
@EnableMockOAuth2Server
@Disabled("")
internal class CorrelationIdFilterTest {
    @Autowired
    private lateinit var securedTestRestTemplate: HttpHeaderTestRestTemplate

    @MockkBean(relaxed = true)
    lateinit var appenderMock: Appender<ILoggingEvent>

    @MockkBean
    lateinit var journalpostServiceMock: JournalpostService

    @LocalServerPort
    private val port = 0

    @BeforeEach
    fun mockLogAppender() {
        val logger = LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME) as Logger
        every { appenderMock.name } returns "MOCK"
        every { appenderMock.isStarted } returns true
        logger.addAppender(appenderMock)
    }

    @Test
    @DisplayName("skal logge requests mot applikasjonen")
    fun skalLoggeRequestsMotApplikasjonen() {
        every {
            journalpostServiceMock.hentJournalpost(
                any(),
                any<KildesystemIdenfikator>(),
            )
        } returns HttpResponse.from(HttpStatus.I_AM_A_TEAPOT)
        val response =
            securedTestRestTemplate.getForEntity<JournalpostResponse>("http://localhost:$port/bidrag-dokument/journal/BID-123?saksnummer=777")
        assertSoftly {
            Assertions.assertThat(response).extracting { it.statusCode }
                .isEqualTo(HttpStatus.I_AM_A_TEAPOT);
            {
                val loggingEventCaptor = ArgumentCaptor.forClass(ILoggingEvent::class.java)
                Mockito.verify(appenderMock, Mockito.atLeastOnce())
                    .doAppend(loggingEventCaptor.capture())
            }
        }
    }
}
